File: internal/handlers/currency.go
package handlers

import (
	"net/http"
	"strconv"
	"worklio-api/internal/services"

	"github.com/labstack/echo/v4"
)

type CurrencyHandler struct {
	exchangeService *services.ExchangeRateService
}

func NewCurrencyHandler(exchangeService *services.ExchangeRateService) *CurrencyHandler {
	return &CurrencyHandler{
		exchangeService: exchangeService,
	}
}

// SupportedCurrency represents a supported currency
type SupportedCurrency struct {
	Code   string `json:"code"`
	Symbol string `json:"symbol"`
	Name   string `json:"name"`
}

// GetSupportedCurrencies godoc
// @Summary Get supported currencies
// @Description Returns a list of all supported currencies
// @Tags currency
// @Produce json
// @Success 200 {object} []SupportedCurrency
// @Router /api/supported-currencies [get]
func (h *CurrencyHandler) GetSupportedCurrencies(c echo.Context) error {
	currencies := []SupportedCurrency{
		{Code: "USD", Symbol: "$", Name: "US Dollar"},
		{Code: "EUR", Symbol: "€", Name: "Euro"},
		{Code: "GBP", Symbol: "£", Name: "British Pound"},
		{Code: "JPY", Symbol: "¥", Name: "Japanese Yen"},
		{Code: "AUD", Symbol: "A$", Name: "Australian Dollar"},
		{Code: "CAD", Symbol: "C$", Name: "Canadian Dollar"},
		{Code: "CHF", Symbol: "CHF", Name: "Swiss Franc"},
		{Code: "CNY", Symbol: "¥", Name: "Chinese Yuan"},
		{Code: "SEK", Symbol: "kr", Name: "Swedish Krona"},
		{Code: "NZD", Symbol: "NZ$", Name: "New Zealand Dollar"},
		{Code: "IDR", Symbol: "Rp", Name: "Indonesian Rupiah"},
		{Code: "SGD", Symbol: "S$", Name: "Singapore Dollar"},
		{Code: "INR", Symbol: "₹", Name: "Indian Rupee"},
	}

	// Verify all currencies are in the service's supported list
	supportedMap := make(map[string]bool)
	for _, code := range services.SupportedCurrencies {
		supportedMap[code] = true
	}

	// Filter to only return currencies that are actually supported
	var result []SupportedCurrency
	for _, currency := range currencies {
		if supportedMap[currency.Code] {
			result = append(result, currency)
		}
	}

	return c.JSON(http.StatusOK, result)
}

// ConvertCurrencyRequest represents a currency conversion request
type ConvertCurrencyRequest struct {
	Amount   float64 `json:"amount" validate:"required"`
	From     string  `json:"from" validate:"required"`
	To       string  `json:"to" validate:"required"`
}

// ConvertCurrencyResponse represents a currency conversion response
type ConvertCurrencyResponse struct {
	Amount         float64 `json:"amount"`
	From           string  `json:"from"`
	To             string  `json:"to"`
	ConvertedAmount float64 `json:"converted_amount"`
	Rate           float64 `json:"rate"`
}

// ConvertCurrency godoc
// @Summary Convert amount between currencies
// @Description Converts an amount from one currency to another using current exchange rates
// @Tags currency
// @Accept json
// @Produce json
// @Param request query ConvertCurrencyRequest true "Conversion Request"
// @Success 200 {object} ConvertCurrencyResponse
// @Failure 400 {object} models.ErrorResponse
// @Failure 500 {object} models.ErrorResponse
// @Router /api/convert-currency [get]
func (h *CurrencyHandler) ConvertCurrency(c echo.Context) error {
	amountStr := c.QueryParam("amount")
	from := c.QueryParam("from")
	to := c.QueryParam("to")

	if amountStr == "" || from == "" || to == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Missing required parameters"})
	}

	amount, err := strconv.ParseFloat(amountStr, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Invalid amount"})
	}

	convertedAmount, err := h.exchangeService.ConvertAmount(c.Request().Context(), amount, from, to)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	// Calculate the rate
	rate := 1.0
	if amount > 0 {
		rate = convertedAmount / amount
	}

	return c.JSON(http.StatusOK, ConvertCurrencyResponse{
		Amount:          amount,
		From:            from,
		To:              to,
		ConvertedAmount: convertedAmount,
		Rate:            rate,
	})
}

File: web/src/lib/utils/currency.ts
// Currency formatting utilities

export const CURRENCY_SYMBOLS: Record<string, string> = {
	USD: '$',
	EUR: '€',
	GBP: '£',
	JPY: '¥',
	AUD: 'A$',
	CAD: 'C$',
	CHF: 'CHF',
	CNY: '¥',
	SEK: 'kr',
	NZD: 'NZ$',
	IDR: 'Rp',
	SGD: 'S$',
	INR: '₹'
};

export const CURRENCY_NAMES: Record<string, string> = {
	USD: 'USD - US Dollar ($)',
	EUR: 'EUR - Euro (€)',
	GBP: 'GBP - British Pound (£)',
	JPY: 'JPY - Japanese Yen (¥)',
	AUD: 'AUD - Australian Dollar (A$)',
	CAD: 'CAD - Canadian Dollar (C$)',
	CHF: 'CHF - Swiss Franc (CHF)',
	CNY: 'CNY - Chinese Yuan (¥)',
	SEK: 'SEK - Swedish Krona (kr)',
	NZD: 'NZD - New Zealand Dollar (NZ$)',
	IDR: 'IDR - Indonesian Rupiah (Rp)',
	SGD: 'SGD - Singapore Dollar (S$)',
	INR: 'INR - Indian Rupee (₹)'
};

export function getCurrencySymbol(currency: string): string {
	return CURRENCY_SYMBOLS[currency] || '$';
}

function formatNumber(num: number, decimals: number = 2): string {
	return num.toLocaleString('en-US', {
		minimumFractionDigits: decimals,
		maximumFractionDigits: decimals
	});
}

export function formatCurrency(amount: number, currency: string = 'USD'): string {
	const symbol = getCurrencySymbol(currency);
	return `${symbol}${formatNumber(amount, 2)}`;
}

export function formatCurrencyRate(rate: number, currency: string = 'USD'): string {
	const symbol = getCurrencySymbol(currency);
	return `${symbol}${formatNumber(rate, 0)}/hr`;
}
